### PR TITLE
feat: rewrite evaluation improvements 

### LINKED ISSUE #106: feat(rewrite): formal judge rubric selection and score distribution analysis
## Background

The rewrite pipeline's final judge currently uses Option A: **privacy, quality, naturalness** on a 1–10 scale, adapted from an earlier research repo. PR #86 improved the judge prompt wording and iterated on scoring behavior as an interim fix, but the formal rubric decision was never made. This issue closes that out.

Closes #37.

## Options under consideration

| Option | Dimensions | Scale | Notes |
|--------|-----------|-------|-------|
| **A (current)** | privacy, quality, naturalness | 1–10 | Baseline; `quality` and `naturalness` overlap in practice |
| **B** | privacy, utility, naturalness | 1–10 with improved anchors | Renames quality → utility for clarity; adds explicit anchor descriptions per score band |
| **C** | privacy, utility, faithfulness, fluency | 1–5 | More granular separation of reservation (faithfulness) vs. readability (fluency); narrower scale may reduce score clustering |

## Scope of work

- [ ] Run score distribution analysis on TAB and NVIDIA Synthetic Biographies using current rubric (Option A) — check for score clustering, ceiling effects, and correlation between dimensions
- [ ] Evaluate Options B and C against the same datasets and compare differentiation
- [ ] Document chosen rubric with justification and sample score distributions
- [ ] Update anchor point descriptions for chosen option
- [ ] Update `FinalJudgeWorkflow` implementation
- [ ] Update test coverage

## Related
- Supersedes #37
- See #98 for LLM-as-a-judge coverage for REPLACE mode

### PR DESCRIPTION
**Pull request by memadi-nv**
# PR Summary — Rewrite Evaluation

## Overview

Separates rewrite evaluation from `run()` / `preview()` and exposes it as an explicit `Anonymizer.evaluate()` call, consistent with replace evaluation. Adds detection validity scoring to rewrite evaluation, replaces numeric judge scores with categorical rubrics, and renames the fluency dimension.

---

## Core Changes

- **`Anonymizer.evaluate()` now supports rewrite results** — `AnonymizerResult` and `PreviewResult` carry `rewrite_config` so `evaluate()` can dispatch without the user restating the mode. `evaluate()` on a rewrite result validates only the evaluate model aliases it actually uses (`check_rewrite=False`), not the full rewrite pipeline roles.

- **Final judge moved out of `run()` / `preview()`** — the repair loop (leakage, utility, `needs_human_review`) still runs as part of every `run()` call. The holistic quality judge runs only when `evaluate()` is called explicitly.

- **Detection validity added to rewrite evaluation** — `detection_valid` is a `float | None` in rewrite mode, representing the fraction of detected entities that passed the judge. Passthrough rows (no detected entities) receive `1.0` (trivially valid) with an `INFO` log. If the score cannot be computed, `detection_valid` is `None`, logged as a `WARNING
`, and displayed as **Unavailable**.

- **`rewrite_judge` model role moved to evaluate config** — no longer part of rewrite generation config; now under `selected_models.evaluate.rewrite_judge`, consistent with the other judge roles.

- **Categorical judge scores replace numeric `1–10` rubrics** — the final rewrite judge now returns `low` / `medium` / `high` per dimension, stored as structured JSON under `judge_evaluation`. Higher is better for all three dimensions, rendered as colour-coded badges in `display_record()`.

- **`naturalness` renamed to `style`** — more precisely reflects what the rubric measures (fluency, grammar, readability).

- **`judge_evaluation` column made public** — renamed from `_judge_evaluation`; appears in `display_record()` output and user-facing docs.

- **`_render_scores_section` receives `is_rewrite` as an explicit parameter** — replaces the previous column-name scan that could misfire on user-defined column names.

---
## Display

- `display_record()` for rewrite results renders `detection_valid` alongside utility and leakage scores.
- `detection_valid = None` / `NaN` renders as **Unavailable** in grey rather than a numeric score.
- Categorical judge rubrics (privacy, quality, style) render with `high` / `medium` / `low` badges.

The following shows what rewrite evaluation results look like. Note that when the `Detection Validity` score is below 1.0, the dropdown displays the flagged invalid entities:

Screenshot 2026-06-12 at 10 42 56

---

## Docs and Notebooks

- **`evaluation.md`** — new Rewrite Evaluation section; detection validity special-value tables for both modes including no-entities and unavailable scenarios with their log messages.
- **`rewrite.md`** — output columns table distinguishes `run()` vs `evaluate()` columns; new "Evaluating rewrite output" section.
- **Notebooks 04 and 05** — added optional "🔬 Evaluate" section with `anonymizer.evaluate(result)` and `evaluated.display_record(0)` as separate cells; Python sources synced.
---

## Tests

- `_standard_side_effect` trimmed to 2 results (pipeline + evaluate); `stub_judge_df` scoped to judge-produced columns only.
- New `_detection_valid_fraction` tests covering all branches: `valid=True`, fraction computation, parse failure, `total=0`, `valid=None`.
- New display tests for `detection_valid=None`, `NaN`, and absent column.
- New test asserting `evaluate()` on a rewrite result calls `validate_model_alias_references` with `check_rewrite=False`.

## Type of Change
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Refactoring

## Testing
- [ ] `make test` passes locally
- [ ] `make check` passes locally (format + lint + typecheck + lock-check)
- [ ] Added/updated tests for changes

## Documentation
- [ ] If docs changed: `make docs-build` passes locally

## Related Issues
Closes #106

### CODE REVIEW

File: docs/concepts/rewrite.md [hunk 1]
```diff
@@ -128,16 +128,19 @@ config = AnonymizerConfig(
...
+| `utility_score` | Always | Quality preservation (0.0--1.0). Higher is better. |
+| `leakage_mass` | Always | Weighted privacy leakage. Lower is better. |
+| `weighted_leakage_rate` | Always | Normalized leakage (0.0--1.0) relative to the maximum possible leakage mass. |
+| `any_high_leaked` | Always | Whether any high-sensitivity entity leaked through. |
+| `needs_human_review` | Always | Flag for records that may need manual review. |
+| `detection_valid` | After `evaluate()` | Fraction of detected entities that passed the detection judge (0.0--1.0); `None` if judge unavailable. |
+| `detection_invalid_entities` | After `evaluate()` | Flagged detections with value, label, and one-sentence reasoning. |
+| `judge_evaluation` | After `evaluate()` | Dict with `privacy`, `quality`, and `style` rubric scores and reasoning. |
 
-Use `preview.trace_dataframe` for the full pipeline trace (domain, disposition, QA pairs, repair iterations, judge evaluation).
```
Comment (memadi-nv) [inline docs/concepts/rewrite.md:140]: Since `judge_evaluation` is explicitly rendered in display_record(), I think it belongs in the same public-facing tier as `utility_score`, `leakage_mass`, and `detection_valid` — all without the underscore prefix- hence removed the internal use, aka `_`

File: src/anonymizer/config/default_model_configs/evaluate.yaml [hunk 2]
```diff
@@ -6,7 +6,13 @@
...
+  # --- Shared ---
   detection_validity_judge: gpt-oss-120b
+
+  # --- Replace evaluation ---
   replace_type_fidelity_judge: gpt-oss-120b
   replace_relational_consistency_judge: gpt-oss-120b
   replace_attribute_fidelity_judge: gpt-oss-120b
+
+  # --- Rewrite evaluation ---
+  rewrite_judge: nemotron-30b-thinking
```
Comment (memadi-nv) [inline src/anonymizer/config/default_model_configs/evaluate.yaml:18]: The `rewrite judge` is one holistic call because the dimensions are entangled. Privacy, quality, and style in a rewrite are hard to evaluate in isolation in contrast with replace judges.

File: src/anonymizer/engine/rewrite/rewrite_workflow.py [hunk 3]
```diff
@@ -381,40 +406,88 @@ def _run_evaluate_repair_loop(
...
+        all_failed: list[FailedRecord] = []
+
+        # --- Detection validity judge ---
+        detection_result = self._detection_judge_wf.evaluate(
+            entity_rows,
+            model_configs=model_configs,
+            selected_models=selected_models,
+            preview_num_records=preview_num_records,
+        )
+        entity_rows = _join_new_columns(entity_rows, detection_result.dataframe)
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/rewrite/rewrite_workflow.py]: `_join_new_columns()` treats row loss as a critical workflow failure and filters the target down to surviving record IDs. Since Anonymizer.evaluate() is post-hoc/diagnostic, a judge timeout or dropped row should not remove a previously successful rewrite row from `evaluated.dataframe`.

Can we use a preserving left join with defaults as described at L147-L150?

File: src/anonymizer/engine/rewrite/rewrite_workflow.py [hunk 4]
```diff
@@ -47,11 +51,42 @@
...
+        logger.warning(
+            "Could not parse entities_by_value to compute detection_valid fraction; defaulting to None.",
+            exc_info=True,
+        )
+        return None
+    if total == 0:
+        # Reachable only when valid is False (True returns early above).
+        # Judge flagged invalid detections but no entities were found — score is uncomputable.
+        return None
+    return max(0.0, (total - invalid_count) / total)
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/rewrite/rewrite_workflow.py:89]: There is an inconsistent judge-output case here. If the detection judge returns `all_valid=False` but `invalid_entities=[]`, this computes `(total - 0) / total == 1.0`, so rewrite evaluation reports perfect detection validity even though the judge explicitly failed the row. The replace display path handles this contradictory case as not satisfied, but after converting the verdict to a fraction we lose that signal.

Maybe return `None` for this case, or otherwise avoid surfacing `1.0`, and add a small regression test for `detection_valid=False` with an empty `detection_invalid_entities` list.

**Review by lipikaramaswamy — COMMENTED** [review]
One bundled-skill follow-up: should we update `skills/anonymizer/SKILL.md` as part of this PR?

This PR changes `Anonymizer.evaluate(result)` so rewrite outputs can also be evaluated post-hoc, but the bundled anonymizer skill still describes evaluation as Replace-only and its generated script only runs evaluation under `if args.evaluate and config.replace is not None:`. It also lists the evaluate model roles without `rewrite_judge`.

Since this is a public workflow/API change, the bundled skill should be kept in sync so agents using it don’t skip the new rewrite evaluation path.

File: src/anonymizer/engine/rewrite/rewrite_workflow.py [hunk 5]
```diff
@@ -381,40 +425,92 @@ def _run_evaluate_repair_loop(
...
+        passthrough_rows[COL_JUDGE_EVALUATION] = None
+
+        if entity_rows.empty:
+            combined = merge_and_reorder(passthrough_rows)
+            return RewriteResult(dataframe=combined, failed_records=[])
+
+        all_failed: list[FailedRecord] = []
+
+        # --- Detection validity judge ---
+        detection_result = self._detection_judge_wf.evaluate(
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/rewrite/rewrite_workflow.py]: Could we make this detection-judge step non-fatal as well? It's outside a `try` so if it fails entirely, `.evaluate()` raises instead of returning `detection_valid=None` and `detection_invalid_entities=[]`.

File: skills/anonymizer/SKILL.md
Comment (lipikaramaswamy) [inline skills/anonymizer/SKILL.md]: This is out of date - still says “scores Replace output”, though the template now evaluates both modes

File: skills/anonymizer/SKILL.md [hunk 6]
```diff
@@ -43,7 +43,7 @@ Read `workflows/interactive.md` and follow it. Anonymization is high-stakes —
...
+- **The evaluation judges use their own model roles** (`detection_validity_judge`, `replace_type_fidelity_judge`, `replace_relational_consistency_judge`, `replace_attribute_fidelity_judge`, `rewrite_judge`), configured in the `evaluate` section of `models.yaml`. They are **not** consumed by `preview()` / `run()`, so a config that anonymizes fine can still fail validation at `evaluate()` if those roles are unset. Defaults ship in `src/anonymizer/config/default_model_configs/evaluate.yaml`.
 - **`*_valid` verdict columns are `True` / `False` / `None`.** `None` means the judge was unavailable (model/infra failure), **not** that the row passed — treat it as "unscored", never as a pass. Inspect verdicts per record with `evaluated.display_record(i)`.
```
Comment (lipikaramaswamy) [inline skills/anonymizer/SKILL.md]: This says `*_valid` columns are booleans, but rewrite `detection_valid` is a float fraction.

File: docs/concepts/rewrite.md [hunk 7]
```diff
@@ -11,7 +11,7 @@ Instead of replacing individual entities, rewrite mode transforms the entire tex
...
 
-The text is then rewritten to reduce identifiability, applying targeted transformations that disrupt inference (e.g., weakening or removing linking details) rather than simply rewording content. The rewritten output is evaluated for both quality and privacy leakage using adversarial testing. If thresholds are exceeded, the system automatically refines the rewrite. A final judge provides a qualitative assessment of the rewritten record. Any records that failed to meet standards are tagged for human review. 
+The text is then rewritten to reduce identifiability, applying targeted transformations that disrupt inference (e.g., weakening or removing linking details) rather than simply rewording content. The rewritten output is evaluated for both quality and privacy leakage using adversarial testing. If thresholds are exceeded, the system automatically refines the rewrite. Any records that failed to meet standards are tagged for human review.
```
Comment (lipikaramaswamy) [inline docs/concepts/rewrite.md]:
```suggestion
The text is then rewritten to reduce identifiability, applying targeted transformations that disrupt inference (e.g., weakening or removing linking details) rather than simply rewording content. The rewritten output is evaluated for both quality and privacy leakage using adversarial testing. If thresholds are exceeded, the system automatically refines the rewrite. Any records that failed to meet standards are tagged for human review.
```
remove trailing space

File: src/anonymizer/interface/display.py [hunk 8]
```diff
@@ -413,63 +413,132 @@ def _normalize_replacement_map(raw: str | dict | object) -> list[dict[str, str]]
...
-        parts.append(f"<span style='margin-right:16px'><strong>Leakage:</strong> {leakage:.2f}</span>")
+        metric_parts.append(f"<span style='margin-right:16px'><strong>Leakage:</strong> {leakage:.2f}</span>")
     if weighted_leakage_rate is not None:
-        parts.append(
+        metric_parts.append(
             "<span style='margin-right:16px'><strong>Weighted Leakage Rate:</strong> "
             f"{weighted_leakage_rate:.2f}</span>"
         )
     if needs_review is not None:
+        label = "Rewrite Need Review" if is_rewrite else "Needs Review"
```
Comment (lipikaramaswamy) [inline src/anonymizer/interface/display.py]:
```suggestion
        label = "Rewrite Needs Review" if is_rewrite else "Needs Review"
```

File: tests/interface/test_display.py [hunk 9]
```diff
@@ -665,3 +667,117 @@ def test_render_record_html_replace_mode_unchanged_when_no_rewritten_column() ->
...
+            "text": "Alice works at Acme",
+            "text_rewritten": "Beth works at Globex",
+            COL_DETECTED_ENTITIES: {"entities": []},
+            "utility_score": 0.9,
+            "leakage_mass": 0.1,
+            "needs_human_review": True,
+        }
+    )
+    result = render_record_html(row, record_index=0)
+    assert "Rewrite Need Review" in result
```
Comment (lipikaramaswamy) [inline tests/interface/test_display.py]:
```suggestion
    assert "Rewrite Needs Review" in result
```

**Review by lipikaramaswamy — APPROVED** [review]
Looks good, you'll have to rebase and handle the merge conflict.  The conflict looks limited to the model-loader import in `tests/interface/test_anonymizer_interface.py`; the resolution should keep both `load_default_model_providers` and `validate_model_alias_references`.
